mlir/include/mlir/IR/BuiltinAttributes.td

```diff
@@ -701,8 +701,10 @@ def Builtin_IntegerAttr : Builtin_Attr<"Integer", "integer",
         return $_get(type.getContext(), type, apValue);
       }
 
+      // TODO: Avoid implicit trunc?
       IntegerType intTy = ::llvm::cast<IntegerType>(type);
-      APInt apValue(intTy.getWidth(), value, intTy.isSignedInteger());
+      APInt apValue(intTy.getWidth(), value, intTy.isSignedInteger(),
+                    /*implicitTrunc=*/true);
```

Comment:
It seems strange we hit an issue here since we're passing the expected `isSigned` here right?

Reply:
Yes, places marked with TODO *probably* shouldn't be using implicitTrunc. Here are the failures if we drop it: https://gist.github.com/nikic/d69e30cf1d28ef5988363dc11e203159

I'm guessing the main problem is https://github.com/llvm/llvm-project/blob/0e8555d4dbfdfeddc01dc2ecf9a9b6e804f7b645/mlir/lib/Conversion/IndexToSPIRV/IndexToSPIRV.cpp#L120 trying to construct -1 of either an unsigned or signless integer type.

Reply:
What is the right way to avoid an implicit truncation with signless integers?

Reply:
Not familiar with MLIR, but probably by using a ctor that accepts APInt? That way you can explicitly specify that the constant needs to be sign extended.

Or possibly the code here should be setting signed=true for signless integers, as parameter is int64_t so signed value should be the default assumption?

Reply:
> Or possibly the code here should be setting signed=true for signless integers, as parameter is int64_t so signed value should be the default assumption?

Tried that, and it causes other failures. Unsurprisingly, there seems to be a mix of assumptions on how to interpret a signless integer constant :)

Reply:
The "proper" solution here is probably to either a) Only allow constructing signless integers from APInt or b) only allow constructing them from plain integers for bit widths <= 64 bit, as the sign distinction only becomes really problematic for larger bit widths.

But in any case, this is not something I want to touch in this PR...

Reply:
Seems like you merged without further approvals or resolving completely this thread here?

Reply:
Sorry, I took this as a discussion on how the TODO could be resolved in the future, not a blocking concern for this PR. Was there something you wanted me to change in *this* PR?

Reply:
> // TODO: Avoid implicit trunc?

This isn't super clear or inciting to a fix. The TODO is a question? Such TODO looks to me like a PR "not ready to merge" because the actual "TODO" needs to be figured.

I would have tried encoded more of the discussion in the TODO actually, to make it more accessible for anyone seeing this TODO in terms of investigating what to do to fix it and not have to rediscover the investigation you done here.
I would think that a better TODO would look something like:

> // TODO: We shouldn't use implicit trunc here, at the moment however treating signless integer creation .....

I'm  actually not even sure how to finish the sentence, you didn't expand enough on the problem you saw with signless integer ("Tried that, and it causes other failures.").

Reply:
I see. I agree that the TODO is not super clear without some context. What I'd like to do is to add a reference to https://github.com/llvm/llvm-project/issues/112510 (which I just filed) to these TODOs, so there is a place that provides a more detailed explanation of the purpose of the TODO and how it may be resolved.

I don't want to add detailed analysis to individual TODO comments, as the point of leaving them is so that I *don't* have to analyze each one in detail. (For the record, I'm trying to get this assertion enabled for more than half a year already, and not tracking down everything to its leafs is part of the compromise to make this feasible at all.)